### PR TITLE
Add ssm document capability

### PR DESCRIPTION
**Pull request by HarrisonHammond**
- Add ssm document capability
- Updated readme

SSM documents can now be run on instances without having to enter the instance. Only persons with permissions for the document will be able to execute the file on the instance.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,7 +9,7 @@ Usage:
 
 /usr/local/bin/aws-connect -a ssh|tunnel [-n <instance name>|-t <instance tag>] [-r region] [-p profile name] [-o port] [-x instance id] [-s] [-h] [-v]
 
-  -a   Connect interactive session or establish ssh tunnel (default: ssh)
+  -a   Connect interactive session, establish ssh tunnel, or run an ssm document on an instance (default: ssh)
```
Comment (dnorth98) [inline README.md]: We should probably add the options here.  ie

>Connect interactive session (ssh), establish tunnel (tunnel), or run an ssm document (document) on an instance (default: ssh)

File: README.md [hunk 2]
```diff
@@ -18,6 +18,9 @@ Usage:
   -x   override Name tag and connect direct to given instance ID
   -s   Pick a specific instance ID
   -h   Display this help
+  -d   Specify the name of the ssm document to run
+  -c   Values for the ssm document parameters
```
Comment (dnorth98) [inline README.md]: Signal they are optional

> -c   Values for the ssm document arguments (if any are required)

File: README.md [hunk 3]
```diff
@@ -18,6 +18,9 @@ Usage:
...
+  -g   The location in ssm of the github token to use
```
Comment (dnorth98) [inline README.md]: > The location in ssm parameter store of a github personal access token to use (required for private repos)

As I wrote this I thought of something.  Given this is in aws-connect, which is public, we should support scripts in PUBLIC repos too.  In this case, we won't require a PAT.

Can you code/test that case too?  Just try some scripts in your own repo.

File: README.md [hunk 4]
```diff
@@ -55,3 +58,6 @@ The SSH tunnel can then be used for things like connecting to an RDS database th
 
 `aws-connect -s -t CLUSTER=prod`
 
+5. Run SSM Document named shell-script on shopify with default profile and parameters 'param1 param 2'. Github token and document are required:
```
Comment (dnorth98) [inline README.md]: change parameters to arguments.  As mentioned above, let's make github token optional if the scripts are in a public repo

File: README.md [hunk 5]
```diff
@@ -55,3 +58,6 @@ The SSH tunnel can then be used for things like connecting to an RDS database th
 
 `aws-connect -s -t CLUSTER=prod`
 
+5. Run SSM Document named shell-script on shopify with default profile and parameters 'param1 param 2'. Github token and document are required: 
+
+`aws-connect -x instance -r region -p default -a document -d shell-script -p default -c 'param1 "param 2"' -g <token_location>`
```
Comment (dnorth98) [inline README.md]: When giving an actual example command, don't use `<token_location>` but rather given an example location.  ie. `/devops/MY_GITHUB_PAT`

File: aws-connect [hunk 6]
```diff
@@ -19,9 +19,12 @@ version() {
 usage() {
   echo "Usage:"
   echo ""
-  echo "$programname -a ssh|tunnel [-n <instance name>|-t <instance tag>] [-r region] [-p profile name] [-o port] [-x instance id] [-s] [-h] [-v]"
+  echo "$programname -a ssh|tunnel|document [-d document name] [-c document parameters] [-g github access token location] [-n <instance name>|-t <instance tag>] [-r region] [-p profile name] [-o port] [-x instance id] [-s] [-h] [-v]"
```
Comment (dnorth98) [inline aws-connect]: Consistency (which is my bad).  Google for the "standard" for putting args in the help and make them standard across all of these (ie. should we have `<>` or not?

File: aws-connect [hunk 7]
```diff
@@ -19,9 +19,12 @@ version() {
...
   echo ""
   echo "  -a   Connect interactive session or establish ssh tunnel (default: ssh)"
+  echo "  -d   Run an ssm document on a specified EC2 instance"
+  echo "  -c   Optional parameters if required to run an ssm document"
```
Comment (dnorth98) [inline aws-connect]: arguments

File: aws-connect [hunk 8]
```diff
@@ -208,6 +220,55 @@ elif [ "${action}" == "tunnel" ]; then
     --parameters "{\"portNumber\":[\"22\"],\"localPortNumber\":[\"${port}\"]}" \
     --region "${aws_region}" \
     --profile "${aws_profile}"
+elif [ "${action}" == "document" ]; then
+
+  if [ -z "${github_token_location}" ]
+  then
+    echo "No github token"
```
Comment (dnorth98) [inline aws-connect]: Feels more like debug.  If leaving in change to something like

```
echo "No github token provided - scripts referenced in SSM document must be in a public repo"
```

File: aws-connect [hunk 9]
```diff
@@ -208,6 +220,55 @@ elif [ "${action}" == "tunnel" ]; then
...
+    document_parameters="parameters='${document_parameters}'"
+  else
+    document_parameters="githubTokenLocation='${github_token_location}',parameters='${document_parameters}'"
+    echo "Github token found"
```
Comment (dnorth98) [inline aws-connect]: Remove `echo "Github token found"`

File: aws-connect [hunk 10]
```diff
@@ -208,6 +220,55 @@ elif [ "${action}" == "tunnel" ]; then
...
+    --parameters "${document_parameters}" \
+    --region ${aws_region} \
+    --profile ${aws_profile} \
+    --query "Command.CommandId") 
+
+  output=$(aws ssm list-command-invocations \
+    --command-id ${sh_command_id} \
+    --details)
+
+  status=$(echo $output | jq -r '.[]' | jq -r '.[]' | jq -r '.Status')
```
Comment (dnorth98) [inline aws-connect]: Don't use jq here.  jq doesn't come standard on the mac.  Use the `--query` option to aws cli.

https://docs.aws.amazon.com/cli/latest/userguide/cli-usage-output.html#cli-usage-output-filter

File: aws-connect [hunk 11]
```diff
@@ -208,6 +220,55 @@ elif [ "${action}" == "tunnel" ]; then
...
+  while [ "$status" == "InProgress" ]
+  do
+    echo "Command still running..."
+    sleep 5
+
+    output=$(aws ssm list-command-invocations \
+      --command-id ${sh_command_id} \
+      --details)
+
+    status=$(echo $output | jq -r '.[]' | jq -r '.[]' | jq -r '.Status' )
```
Comment (dnorth98) [inline aws-connect]: same comment...`--query`

File: aws-connect [hunk 12]
```diff
@@ -208,6 +220,55 @@ elif [ "${action}" == "tunnel" ]; then
...
+  done
+
+  if [ "$status" == "Failed" ]
+  then
+    echo "Command Failed!"
+  fi
+
+  echo $output | jq -r '.[]' | jq -r '.[]' | jq -r '.CommandPlugins' | jq -r '.[1]' | jq -r '.Output'
```
Comment (dnorth98) [inline aws-connect]: So I believe the output gets truncated if it's too big for the output field. In that case, it gets written to S3.  Again though here, don't use jq, use query

https://docs.aws.amazon.com/cli/latest/reference/ssm/send-command.html

```
[--output-s3-region <value>]
[--output-s3-bucket-name <value>]
[--output-s3-key-prefix <value>]
```

You'll need to test running a script that produces larger output and see what you get back in output.  I think we may need to pass in an S3 bucket to throw the results into and then return the output location to the user.  The prefix we can auto-generate (say based on the command ID) but if we don't pass a bucket, we'll lose the output

**Review by dnorth98 — CHANGES_REQUESTED** [review]

File: README.md [hunk 13]
```diff
@@ -55,3 +59,10 @@ The SSH tunnel can then be used for things like connecting to an RDS database th
 
 `aws-connect -s -t CLUSTER=prod`
 
+5. Run SSM Document named shell-script on shopify with default profile and arguments 'param1 param 2'. The cloudwatch log name has been changed to ssm-cloudwatch-logs. Document is required, github token is only required for private repos:
```
Comment (dnorth98) [inline README.md]: Shopify?  aws-connect knows nothing about Shopify...

File: README.md [hunk 14]
```diff
@@ -55,3 +59,10 @@ The SSH tunnel can then be used for things like connecting to an RDS database th
 
 `aws-connect -s -t CLUSTER=prod`
 
+5. Run SSM Document named shell-script on shopify with default profile and arguments 'param1 param 2'. The cloudwatch log name has been changed to ssm-cloudwatch-logs. Document is required, github token is only required for private repos: 
+
+`aws-connect -x instance -r region -p default -a document -d shell-script -p default -w 'param1 "param 2"' -g /devops/github_token -c ssm-cloudwatch-logs`
```
Comment (dnorth98) [inline README.md]: Use real (made up) values here since it's meant to be an actual example command.  For example

`aws-connect -x i-1234567890 -r us-east-1 -p staging -a document -d my-ssm-document -w 'param1 "param 2"' -g /devops/github_token -c ssm-cloudwatch-logs`

File: README.md [hunk 15]
```diff
@@ -55,3 +59,10 @@ The SSH tunnel can then be used for things like connecting to an RDS database th
 
 `aws-connect -s -t CLUSTER=prod`
 
+5. Run SSM Document named shell-script on shopify with default profile and arguments 'param1 param 2'. The cloudwatch log name has been changed to ssm-cloudwatch-logs. Document is required, github token is only required for private repos: 
+
+`aws-connect -x instance -r region -p default -a document -d shell-script -p default -w 'param1 "param 2"' -g /devops/github_token -c ssm-cloudwatch-logs` 
+
+6. Run SSM Document named shell-script on shopify with default profile and no arguments on a public repo. The cloudwatch log name has been changed to ssm-cloudwatch-logs. Document is required: 
+
+`aws-connect -x instance -r region -p default -a document -d shell-script -p default -c ssm-cloudwatch-logs`
```
Comment (dnorth98) [inline README.md]: Same comment (and -p is here twice)

File: README.md (hunk 13 above)
Comment (dnorth98) [inline README.md]: > github token is only required for private repos:

I think we need to elaborate a bit more here.

> A github token is required if the SSM document is referencing content in a private github repo.  See this [AWS blog post](https://aws.amazon.com/blogs/mt/run-scripts-stored-in-private-or-public-github-repositories-using-amazon-ec2-systems-manager/) on documents referencing private repos

File: README.md [hunk 16]
```diff
@@ -55,3 +59,10 @@ The SSH tunnel can then be used for things like connecting to an RDS database th
 
 `aws-connect -s -t CLUSTER=prod`
 
+5. Run SSM Document named shell-script on shopify with default profile and arguments 'param1 param 2'. The cloudwatch log name has been changed to ssm-cloudwatch-logs. Document is required, github token is only required for private repos: 
+
+`aws-connect -x instance -r region -p default -a document -d shell-script -p default -w 'param1 "param 2"' -g /devops/github_token -c ssm-cloudwatch-logs` 
+
+6. Run SSM Document named shell-script on shopify with default profile and no arguments on a public repo. The cloudwatch log name has been changed to ssm-cloudwatch-logs. Document is required:
```
Comment (dnorth98) [inline README.md]: Document is required but since the document does not access any code in a github private repo, no github token is required.

File: aws-connect [hunk 17]
```diff
@@ -208,6 +224,70 @@ elif [ "${action}" == "tunnel" ]; then
     --parameters "{\"portNumber\":[\"22\"],\"localPortNumber\":[\"${port}\"]}" \
     --region "${aws_region}" \
     --profile "${aws_profile}"
+elif [ "${action}" == "document" ]; then
+
+  if [ -z "${github_token_location}" ]
+  then
+    echo "No github token provided - scripts referenced in SSM document must be in a public repo if no token is provided"
```
Comment (dnorth98) [inline aws-connect]: I feel we can leave this echo out or it will appear for every run if you don't care about github.

File: aws-connect [hunk 18]
```diff
@@ -208,6 +224,70 @@ elif [ "${action}" == "tunnel" ]; then
...
+  else
+    arguments="githubTokenLocation='${github_token_location}',"
+  fi
+
+  if [ -z "${document_parameters}" ]
```
Comment (dnorth98) [inline aws-connect]: general comment.  Some places we use this if syntax, others we use

```
if [ "${action}" == "ssh" ]; then
```

I know I use the single line then format.  I'm not super picky which one we use but keep it consistent within the same script.  multiple styles in the same file make it harder to read.

File: aws-connect [hunk 19]
```diff
@@ -208,6 +224,70 @@ elif [ "${action}" == "tunnel" ]; then
...
+    --command-id "${sh_command_id}" \
+    --details \
+    --output text \
+    --no-paginate \
+    --query "CommandInvocations[0].Status")
+
+  echo "Command ID: $sh_command_id"
+
+  # wait for the command to finish running before getting the output
+  while [ "$status" == "InProgress" ]
```
Comment (dnorth98) [inline aws-connect]: Ah I've been bitten by this before.  If the command never gets to InProgress, you'll never drop into the while loop.  For example, if it starts in `Pending` or `Delayed `, you will never loop.

Check out all the states
https://docs.aws.amazon.com/systems-manager/latest/userguide/monitor-commands.html

I also found that sometimes when SSM is busy, there is just no status!  So loop again if the status is not set

I found a methodology like this worked in another solution I did (this is python) for example

```
ssm_failed_statuses = ['Cancelled', 'TimedOut', 'Failed']
: 
: 
# Do we have a status?
                    if 'CommandInvocations' in command_status:
                        # Check if the command worked
                        if command_status_string == 'Success':
                            instances_command_status[instance] = 'SUCCESS'
                        elif command_status_string in ssm_failed_statuses:
                            status_code = command_status['ResponseCode']
                            instances_command_status[instance] = 'FAILED-' + str(status_code) 
                            failed_commands = True
```

Basically, keep looping until you get a success or a failed status

File: aws-connect [hunk 20]
```diff
@@ -208,6 +224,70 @@ elif [ "${action}" == "tunnel" ]; then
...
+      --no-paginate \
+      --query "CommandInvocations[0].Status")  
+  done
+
+  if [ "$status" == "Failed" ]
+  then
+    echo "Command Failed!"
+  fi
+
+  printf "$(aws logs get-log-events --log-group-name ${cloudwatch_group} --log-stream-name ${sh_command_id}/${instance_id}/runShellScript/stdout --query "events[0].message")"
```
Comment (dnorth98) [inline aws-connect]: 2 comments:

1) For the log stream name, why does it contain `runShellScript`?  Isn't that the name of AWS's SSM document?  It's probably better to include the document name being run AND the user running it (since the log group is shared).  There's a CLI call to get the current IAM user

2) You probably read this, but this is paginated

>By default, this operation returns as many log events as can fit in a response size of 1MB (up to 10,000 log events). You can get additional log events by specifying one of the tokens in a subsequent call.

It's probably ok to just print a line showing the full URL to the log stream for the full logs.  Something like

> Log output is limited to 1MB.  For the full logs, view them at https://console.aws.com....blah blah/log_group/stream (look at an existing for the full url)

**Review by dnorth98 — CHANGES_REQUESTED** [review]

**Review by dnorth98 — APPROVED** [review]
